File: Example/FWFramework/Classes/Module/Test/Module/TestQrcodeController.m
//
//  TestQrcodeController.m
//  FWFramework_Example
//
//  Created by wuyong on 2022/8/25.
//  Copyright © 2022 CocoaPods. All rights reserved.
//

#import "TestQrcodeController.h"
#import "AppSwift.h"
@import FWFramework;

@interface TestQrcodeController () <FWViewController>

@property (nonatomic, strong) FWQrcodeScanManager *scanManager;
@property (nonatomic, strong) FWQrcodeScanView *scanView;

@property (nonatomic, strong) UIButton *flashlightBtn;
@property (nonatomic, assign) BOOL flashlightSelected;
@property (nonatomic, strong) UILabel *promptLabel;

@end

@implementation TestQrcodeController

- (void)viewDidLoad
{
    [super viewDidLoad];
    self.fw_navigationBarStyle = 2;
    self.fw_extendedLayoutEdge = UIRectEdgeTop;
    self.navigationItem.title = @"扫一扫";
    self.navigationItem.rightBarButtonItem = [[UIBarButtonItem alloc] initWithTitle:@"相册" style:UIBarButtonItemStyleDone target:self action:@selector(onPhotoLibrary)];
    
    // 相机授权
    [[FWAuthorizeManager managerWithType:FWAuthorizeTypeCamera] authorize:^(FWAuthorizeStatus status) {
        if (status != FWAuthorizeStatusAuthorized) {
            [self fw_showAlertWithTitle:(status == FWAuthorizeStatusRestricted ? @"未检测到您的摄像头" : @"未打开摄像头权限") message:nil cancel:nil cancelBlock:NULL];
        } else {
            [self setupScanManager];
            [self.view addSubview:self.scanView];
            [self.view addSubview:self.promptLabel];
            
            // 由于异步授权，viewWillAppear时可能未完成，此处调用start
            [self startScanManager];
        }
    }];
}

- (void)viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];
    [self startScanManager];
}

- (void)viewWillDisappear:(BOOL)animated
{
    [super viewWillDisappear:animated];
    [self stopScanManager];
}

- (void)dealloc
{
    [self removeScanView];
}

#pragma mark - Accessor

- (FWQrcodeScanView *)scanView
{
    if (!_scanView) {
        _scanView = [[FWQrcodeScanView alloc] initWithFrame:CGRectMake(0, 0, FWScreenWidth, FWScreenHeight)];
        _scanView.scanImageName = [FWModuleBundle imageNamed:@"qrcodeLine"];
    }
    return _scanView;
}

- (UIButton *)flashlightBtn
{
    if (!_flashlightBtn) {
        _flashlightBtn = [UIButton buttonWithType:UIButtonTypeCustom];
        CGFloat flashlightBtnW = 30;
        CGFloat flashlightBtnH = 30;
        CGFloat flashlightBtnX = 0.5 * (self.view.frame.size.width - flashlightBtnW);
        CGFloat flashlightBtnY = 0.5 * FWScreenHeight + 0.35 * self.view.frame.size.width - flashlightBtnH - 25;
        _flashlightBtn.frame = CGRectMake(flashlightBtnX, flashlightBtnY, flashlightBtnW, flashlightBtnH);
        [_flashlightBtn setBackgroundImage:[FWModuleBundle imageNamed:@"qrcodeFlashlightOpen"] forState:(UIControlStateNormal)];
        [_flashlightBtn setBackgroundImage:[FWModuleBundle imageNamed:@"qrcodeFlashlightClose"] forState:(UIControlStateSelected)];
        [_flashlightBtn addTarget:self action:@selector(toggleFlashlightBtn:) forControlEvents:UIControlEventTouchUpInside];
    }
    return _flashlightBtn;
}

- (UILabel *)promptLabel
{
    if (!_promptLabel) {
        CGFloat tipLabelY = 0.5 * FWScreenHeight + 0.35 * self.view.frame.size.width + 12;
        _promptLabel = [[UILabel alloc] initWithFrame:CGRectMake(0, tipLabelY, FWScreenWidth, 20)];
        _promptLabel.font = [UIFont systemFontOfSize:13];
        _promptLabel.textColor = [AppTheme textColor];
        _promptLabel.textAlignment = NSTextAlignmentCenter;
        _promptLabel.text = @"将二维码/条码放入框内, 即可自动扫描";
    }
    return _promptLabel;
}

#pragma mark - Private

- (void)setupScanManager
{
    self.scanManager = [[FWQrcodeScanManager alloc] init];
    self.scanManager.sampleBufferDelegate = YES;
    [self.scanManager scanQrcodeWithView:self.view];
    
    FWWeakifySelf();
    self.scanManager.scanResultBlock = ^(NSString *result) {
        FWStrongifySelf();
        if (result) {
            [UIApplication fw_playSystemSound:[FWModuleBundle resourcePath:@"Qrcode.caf"]];
            [self stopScanManager];
            
            [self onScanResult:result];
        }
    };
    self.scanManager.scanBrightnessBlock = ^(CGFloat brightness) {
        FWStrongifySelf();
        if (brightness < -1) {
            [self.view addSubview:self.flashlightBtn];
        } else {
            if (self.flashlightSelected == NO) {
                [self removeFlashlightBtn];
            }
        }
    };
}

- (void)startScanManager
{
    [self.scanManager startRunning];
    [self.scanView addTimer];
}

- (void)stopScanManager
{
    [self.scanView removeTimer];
    [self removeFlashlightBtn];
    [self.scanManager stopRunning];
}

- (void)toggleFlashlightBtn:(UIButton *)button
{
    if (button.selected == NO) {
        [FWQrcodeScanManager openFlashlight];
        
        self.flashlightSelected = YES;
        button.selected = YES;
    } else {
        [self removeFlashlightBtn];
    }
}

- (void)removeFlashlightBtn
{
    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.2 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
        [FWQrcodeScanManager closeFlashlight];
        
        self.flashlightSelected = NO;
        self.flashlightBtn.selected = NO;
        [self.flashlightBtn removeFromSuperview];
    });
}

- (void)removeScanView
{
    [self.scanView removeTimer];
    [self.scanView removeFromSuperview];
    self.scanView = nil;
}

#pragma mark - Action

- (void)onPhotoLibrary
{
    [self stopScanManager];
    
    FWWeakifySelf();
    [self fw_showImagePickerWithFilterType:FWImagePickerFilterTypeImage selectionLimit:1 allowsEditing:NO customBlock:^(UIViewController *imagePicker) {
        FWStrongifySelf();
        if ([imagePicker isKindOfClass:[UIViewController class]]) {
            imagePicker.fw_presentationDidDismiss = ^{
                FWStrongifySelf();
                [self startScanManager];
            };
        }
    } completion:^(NSArray * _Nonnull objects, NSArray * _Nonnull results, BOOL cancel) {
        FWStrongifySelf();
        if (cancel) {
            [self startScanManager];
        } else {
            UIImage *image = objects.firstObject;
            image = [image fw_compressImageWithMaxWidth:1200];
            image = [image fw_compressImageWithMaxLength:300 * 1024];
            NSString *result = [FWQrcodeScanManager scanQrcodeWithImage:image];
            
            [self onScanResult:result];
        }
    }];
}

- (void)onScanResult:(NSString *)result
{
    FWWeakifySelf();
    [self fw_showAlertWithTitle:@"扫描结果" message:FWSafeString(result) cancel:nil cancelBlock:^{
        FWStrongifySelf();
        [self startScanManager];
    }];
}

@end

File: Example/FWFramework/Classes/Module/Test/Module/TestQrcodeController.swift
//
//  TestQrcodeController.swift
//  FWFramework_Example
//
//  Created by wuyong on 2022/9/27.
//  Copyright © 2022 CocoaPods. All rights reserved.
//

import FWFramework

class TestQrcodeController: UIViewController, ViewControllerProtocol {
    private let scanCode = ScanCode()
    private var currentZoomFactor: CGFloat = 1.0

    private lazy var scanView: ScanView = {
        let configuration = ScanViewConfiguration()
        configuration.scanlineImage = ModuleBundle.imageNamed("qrcodeLine")

        let result = ScanView(frame: CGRect(x: 0, y: 0, width: APP.screenWidth, height: APP.screenHeight), configuration: configuration)
        result.scanFrame = CGRect(x: 0, y: 0.18 * APP.screenHeight, width: APP.screenWidth, height: 0.54 * APP.screenHeight)
        result.doubleTapBlock = { [weak self] selected in
            self?.scanCode.videoZoomFactor = selected ? 4.0 : 1.0
        }
        result.pinchScaleBlock = { [weak self] scale in
            guard let self else { return }

            if scale <= 0 {
                currentZoomFactor = scanCode.videoZoomFactor
            } else {
                scanCode.videoZoomFactor = currentZoomFactor * scale
            }
        }
        return result
    }()

    private lazy var flashlightBtn: UIButton = {
        let result = UIButton(type: .custom)
        result.isHidden = true
        result.frame = CGRect(x: (APP.screenWidth - 30) / 2, y: scanView.scanFrame.maxY + 30, width: 30, height: 30)
        result.setBackgroundImage(ModuleBundle.imageNamed("qrcodeFlashlightOpen"), for: .normal)
        result.setBackgroundImage(ModuleBundle.imageNamed("qrcodeFlashlightClose"), for: .selected)
        result.app.addTouch { [weak self] _ in
            self?.toggleFlashlightBtn()
        }
        return result
    }()

    func setupNavbar() {
        let appearance = NavigationBarAppearance()
        appearance.foregroundColor = .white
        appearance.backgroundTransparent = true
        appearance.leftBackImage = Icon.backImage
        app.navigationBarAppearance = appearance

        app.extendedLayoutEdge = .top
        navigationItem.title = "扫一扫"
        app.setRightBarItem(UIBarButtonItem.SystemItem.action.rawValue) { [weak self] _ in
            self?.app.showSheet(title: nil, message: nil, actions: ["相册二维码", "相册条形码", "扫描二维码", "扫描条形码", "同时扫描", "生成二维码", "生成条形码"], actionBlock: { index in
                if index == 0 {
                    self?.onPhotoLibrary(false)
                } else if index == 1 {
                    self?.onPhotoLibrary(true)
                } else if index == 2 {
                    self?.scanCode.metadataObjectTypes = ScanCode.metadataObjectTypesQRCode
                } else if index == 3 {
                    self?.scanCode.metadataObjectTypes = ScanCode.metadataObjectTypesBarcode
                } else if index == 4 {
                    self?.scanCode.metadataObjectTypes = ScanCode.metadataObjectTypesQRCode + ScanCode.metadataObjectTypesBarcode
                } else if index == 5 {
                    self?.onGenerateCode(false)
                } else if index == 6 {
                    self?.onGenerateCode(true)
                }
            })
        }
    }

    func setupSubviews() {
        view.backgroundColor = .black
        view.addSubview(scanView)
        view.addSubview(flashlightBtn)
    }

    override func viewDidLoad() {
        super.viewDidLoad()
        setupScanManager()
    }

    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        startScanManager()
    }

    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)
        stopScanManager()
    }

    func setupScanManager() {
        scanCode.stopWhenScanned = true
        scanCode.scanResultBlock = { [weak self] result in
            if result != nil, let sound = ModuleBundle.resourcePath("Qrcode.caf") {
                ScanCode.playSoundEffect(sound)
            }

            DispatchQueue.app.mainAsync { [weak self] in
                self?.stopScanManager()
                self?.onScanResult(result)
            }
        }
        scanCode.scanBrightnessBlock = { [weak self] brightness in
            if brightness < -1 {
                DispatchQueue.app.mainAsync { [weak self] in
                    self?.flashlightBtn.isHidden = false
                }
            } else {
                if !ScanCode.isTorchActive() {
                    DispatchQueue.app.mainAsync { [weak self] in
                        self?.removeFlashlightBtn()
                    }
                }
            }
        }
        scanCode.preview = view
    }

    func startScanManager() {
        scanCode.startRunning()
        scanView.startScanning()
    }

    func stopScanManager() {
        scanCode.stopRunning()
        scanView.stopScanning()
        removeFlashlightBtn()
    }

    func toggleFlashlightBtn() {
        if !flashlightBtn.isSelected {
            flashlightBtn.isSelected = true

            ScanCode.turnOnTorch()
        } else {
            removeFlashlightBtn()
        }
    }

    func removeFlashlightBtn() {
        ScanCode.turnOffTorch()

        flashlightBtn.isSelected = false
        flashlightBtn.isHidden = true
    }

    func onPhotoLibrary(_ isBarcode: Bool = false) {
        stopScanManager()

        app.showImagePicker(filterType: .image, selectionLimit: 1, allowsEditing: false) { [weak self] imagePicker in
            guard let imagePicker = imagePicker as? UIViewController else { return }
            imagePicker.app.presentationDidDismiss = {
                self?.startScanManager()
            }
        } completion: { [weak self] objects, _, cancel in
            if cancel {
                self?.startScanManager()
            } else {
                if !isBarcode {
                    self?.app.showLoading(text: "识别中...")
                    ScanCode.readQRCode(objects.first as? UIImage) { result in
                        self?.app.hideLoading()

                        self?.onScanResult(result)
                    }
                } else {
                    self?.app.showLoading(text: "识别中...")
                    ScanCode.readBarcode(objects.first as? UIImage) { result in
                        self?.app.hideLoading()

                        self?.onScanResult(result)
                    }
                }
            }
        }
    }

    func onGenerateCode(_ isBarcode: Bool = false) {
        stopScanManager()

        app.showPrompt(title: nil, message: "Please input code", cancel: nil, confirm: nil, promptBlock: nil) { [weak self] value in
            guard !value.isEmpty else {
                self?.startScanManager()
                return
            }

            var image: UIImage?
            if isBarcode {
                image = ScanCode.generateBarcode(data: value, size: CGSize(width: 355, height: 105))
                image = image?.app.image(insets: UIEdgeInsets(top: -10, left: -10, bottom: -10, right: -10), color: UIColor.white)
            } else {
                image = ScanCode.generateQrcode(data: value, size: 375)
            }
            guard let image else {
                self?.startScanManager()
                return
            }

            self?.app.showImagePreview(imageURLs: [image], imageInfos: nil, currentIndex: 0, sourceView: nil)
        } cancelBlock: { [weak self] in
            self?.startScanManager()
        }
    }

    func onScanResult(_ result: String?) {
        if let result {
            app.showAlert(title: "扫描结果", message: result, cancel: nil) { [weak self] in
                self?.startScanManager()
            }
        } else {
            app.showMessage(text: "识别失败")
            startScanManager()
        }
    }
}
